### PR TITLE
Implement an option to not start the config_monitor thread

### PR DESCRIPTION
**Pull request by liuw**
I normally have a lot of terminals open and don't really care if they can reload the latest configuration or not.

It seems that I'm not alone on this, hence this PR.

CC: #670

### CODE REVIEW

**Comment by siiptuo** [comment]
Great work on this useful feature! However I find the naming of `--stop-config-monitor` is a bit confusing because the monitor won't be started. Also it doesn't really align with existing naming conventions. I think that something like `--no-config-monitor`, `--disable-config-monitor`, `--no-config-watch` or `--no-config-refresh` would fit better here.

**Comment by liuw** [comment]
Aye aye. I think any of your suggestions is good. I'm very bad at naming things.

**Comment by lord-re** [comment]
I'm using it right now and it works well.

**Comment by liuw** [comment]
@jwilm Any comment? I updated the option to --no-config-monitor according to @siiptuo 's comment. What else do I need to do to get this pull request merged?

File: src/main.rs [hunk 1]
```diff
@@ -148,8 +148,13 @@ fn run(mut config: Config, options: cli::Options) -> Result<(), Box<Error>> {
     //
     // The monitor watches the config file for changes and reloads it. Pending
     // config changes are processed in the main loop.
-    let config_monitor = config.path()
-        .map(|path| config::Monitor::new(path, display.notifier()));
+    let config_monitor;
```
Comment (NickeZ) [inline src/main.rs]: You can use the following to remove some redundant code.

```
let config_monitor = if !options.no_config_monitor {
...
} else {
...
}
```

Comment (liuw) [inline src/main.rs]: Good point. I will make the change in next iteration.

**Comment by lord-re** [comment]
Maybe adding the (commented) option to the default alacritty.yaml with an explanation ?

**Comment by jwilm** [comment]
> Any comment? I updated the option to --no-config-monitor according to @siiptuo 's comment. What else do I need to do to get this pull request merged?

Just haven't had time to look through PRs recently; sorry!

**Comment by liuw** [comment]
> Maybe adding the (commented) option to the default alacritty.yaml with an explanation ?

There is no new config file option. It is command line only.

**Comment by liuw** [comment]
> Just haven't had time to look through PRs recently; sorry!

No worries. Take your time.

**Comment by DarkDefender** [comment]
I would like if this could be a config option in addition to the command line switch.

**Comment by liuw** [comment]
People asked for a config file option and I finally had the time today to investigate.

Now, suppose you start a few processes with config threads, then you change the config file to disallow config monitor threads. What do you expect the existing threads to do?

There are two options:

1. Allow the existing threads to continue running;
2. Make them exit.

The first option is easy, the second requires a bit of refactoring, so before I go away and write code, I would need to have some opinions first.

**Comment by maximbaz** [comment]
My opinion is that this doesn't really matter much, the second option sounds cleaner, but if it happens that it requires much effort, it is totally fine to require a restart (just like DPI setting does).

I'm curious about another thing, since you will introduce config parameter to disable config monitor, is there a demand for CLI parameter to enable it? 😃 So that users can disable config monitor by default, but when they are about to pick a new color scheme, they would start `$ alacritty --watch-config=yes` on that particular day.

**Comment by jwilm** [comment]
@liuw for this option in config, we should require a restart of Alacritty. Starting/stopping the monitor mid-process doesn't seem worth the additional complexity.

**Comment by jwilm** [comment]
@maximbaz I'm a fan of being able to enable the monitor from command line as well.

**Comment by liuw** [comment]
OK I update the code according to the comments I received. Please review.

File: alacritty.yml [hunk 2]
```diff
@@ -275,6 +275,10 @@ selection:
 
 hide_cursor_when_typing: false
 
+# Start a thread to monitor configuration file changes.
+# (changes require restart)
+start_config_monitor_thread: true
```
Comment (jwilm) [inline alacritty.yml]: Can we name this `live_config_reload` instead of `start_config_monitor_thread`? I think there's no reason to mention how live reloading is implemented.

File: src/cli.rs [hunk 3]
```diff
@@ -59,6 +61,11 @@ impl Options {
             .arg(Arg::with_name("ref-test")
                 .long("ref-test")
                 .help("Generates ref test"))
+            .arg(Arg::with_name("config-monitor-thread")
```
Comment (jwilm) [inline src/cli.rs]: Ditto for the CLI option, can this be `live-config-reload` as well?

File: src/main.rs [hunk 4]
```diff
@@ -148,8 +148,16 @@ fn run(mut config: Config, options: cli::Options) -> Result<(), Box<Error>> {
     //
     // The monitor watches the config file for changes and reloads it. Pending
     // config changes are processed in the main loop.
-    let config_monitor = config.path()
-        .map(|path| config::Monitor::new(path, display.notifier()));
+    let mut config_monitor = None;
```
Comment (jwilm) [inline src/main.rs]: I've got a few nits about this section of code:

1. `config_monitor` is getting wrapped in a second `Option` and ends up with the type `Option<Option<config::Monitor>>`.
2. Code to create the config monitor is not dry (`Monitor::new` called identically in two places).
3. Kind of hard to follow the logic.

I would propose rewriting this as a match expression. That is, something like:

```rust
let config_monitor = match (opts.live_reload, config.live_reload) {
    // Start monitor if CLI flag says yes
    (Some(true), _) |
    // Or if no CLI flag was passed and the config says yes
    (None, true) => (), // create the monitor here
    // Otherwise, don't start the monitor
    _ => None,
};
```

File: src/main.rs [hunk 5]
```diff
@@ -160,15 +168,17 @@ fn run(mut config: Config, options: cli::Options) -> Result<(), Box<Error>> {
...
-        config_monitor.as_ref()
-            .and_then(|monitor| monitor.pending_config())
-            .map(|new_config| {
-                config = new_config;
-                display.update_config(&config);
-                processor.update_config(&config);
-                terminal.update_config(&config);
-                terminal.dirty = true;
-            });
+        if let Some(ref monitor) = config_monitor {
```
Comment (jwilm) [inline src/main.rs]: With the above suggested refactor, this change can go away.

File: alacritty.yml (hunk 2 above)
Comment (liuw) [inline alacritty.yml]: OK, will change.

File: src/cli.rs (hunk 3 above)
Comment (liuw) [inline src/cli.rs]: No problem.

**Comment by jwilm** [comment]
Note that behavior on master is slightly different than what this PR had:

* Config reloading still defaults to enabled to match the original behavior (add a flag to your config file if you want it off by default)
* Passing a command line argument now uses flags. Here's the relevant help:

      --live-config-reload       Enable automatic config reloading
      --no-live-config-reload    Disable automatic config reloading

For reasoning on the second change, please see the commit message of 59c1dbf

**Comment by jwilm** [comment]
Thanks for the work on this feature @liuw!
